### PR TITLE
Bump Acorn JS version to ES2018

### LINKED ISSUE #11431: acorn-optimizer.js trips over async functions
When using `async function`s in the pre or post JS and combining that with `-s MAXIMUM_MEMORY=4GB` the `acorn-optimizer.js` trips over the `async function` during parsing. Looks like acorn could use an update :-)

### PR DESCRIPTION
**Pull request by kripken**
The acorn code says
```javascript
  // `ecmaVersion` indicates the ECMAScript version to parse. Must be
  // either 3, 5, 6 (2015), 7 (2016), 8 (2017), 9 (2018), or 10
  // (2019). This influences support for strict mode, the set of
  // reserved words, and support for new syntax features. The default
  // is 9.
```
We set the version to 6 (2015) when we added Acorn, probably to
keep it as close as possible to the old pre-Acorn parser. But the
default in our version is 9 (2018) so we may as well switch to that,
and that fixes #11431 (8 - 2017 - would have been enough for
that specific bug).

### CODE REVIEW

**Comment by RReverser** [comment]
Sounds good; we at Acorn are thinking of bumping the `ecmaVersion` default in the next major release anyway, but it's best to set it explicitly anyway.

File: tools/acorn-optimizer.js [hunk 1]
```diff
@@ -1299,7 +1299,7 @@ if (extraInfoStart > 0) {
 // Collect all JS code comments to this array so that we can retain them in the outputted code
 // if --closureFriendly was requested.
 var sourceComments = [];
-var ast = acorn.parse(input, { ecmaVersion: 6, preserveParens: closureFriendly, onComment: closureFriendly ? sourceComments : undefined });
+var ast = acorn.parse(input, { ecmaVersion: 9, preserveParens: closureFriendly, onComment: closureFriendly ? sourceComments : undefined });
```
Comment (RReverser) [inline tools/acorn-optimizer.js:1302]: I'd probably use modern version notation instead; pretty much nobody says "ES9" but rather "ES2018".

```suggestion
var ast = acorn.parse(input, { ecmaVersion: 2018, preserveParens: closureFriendly, onComment: closureFriendly ? sourceComments : undefined });
```

Comment (RReverser) [reply tools/acorn-optimizer.js:1302]: Or might even switch to `2019` right away.

```suggestion
var ast = acorn.parse(input, { ecmaVersion: 2019, preserveParens: closureFriendly, onComment: closureFriendly ? sourceComments : undefined });
```

**Comment by kripken** [comment]
Thanks! I didn't know `2018` was an accepted syntax.

I think `2018` is better than `2019` for now, as I'm not sure if `2019` was stable when we branched our version of Acorn. We should probably update.

**Comment by RReverser** [comment]
> Thanks! I didn't know 2018 was an accepted syntax.

It says so in the comment you quoted, but maybe we didn't make it clear enough that versions in parentheses are aliases :)

**Comment by RReverser** [comment]
> I think 2018 is better than 2019 for now, as I'm not sure if 2019 was stable when we branched our version of Acorn. 

FWIW core Acorn doesn't implement any features that haven't been finalised yet (< stage 4), so you should be safe either way; upgrading sounds like a good idea though, especially since the latest version has stable `2020` support.

**Comment by kripken** [comment]
Heh, I should read the full text of things I quote I guess :smile:

**Comment by sbc100** [comment]
On the subject of acorn, what is are blockers on us upstreaming out local patches and switching to the npm version?   Would be nice to avoid the local copy.

**Comment by kripken** [comment]
I believe they just didn't want one of our patches, https://github.com/acornjs/acorn/pull/793 It might be possible to do something non-invasive that achieves the same result, but I don't know.

**Comment by RReverser** [comment]
> It might be possible to do something non-invasive that achieves the same result, but I don't know.

Yeah the error object thrown by Acorn contains location in the source it was thrown at, so you don't really need a patch, only a wrapper to try-catch it and format it in whichever way you prefer.

**Comment by sbc100** [comment]
I've love to see this happen... I'll open bug and hopefully one of us can get around to it: #11439

**Pull request merged**